### DISCUSSION TITLE
urql /w typescript throws error on creating mock client

### DISCUSSION THREAD

**Discussion by basvandriel** [urql-graphql/urql#2057] [Q&A] [answered]
When creating a mock client for testing components in TypeScript it throws the following error:

```
Type '{ executeMutation: Mock<any, any>; }' is missing the following properties from type 'Client': operations$, reexecuteOperation, url, suspense, and 12 more.ts(2740)
index.d.ts(335, 9): The expected type comes from property 'value' which is declared here on type 'IntrinsicAttributes & ProviderProps<Client>'
```

Using the following code:

```
it("should do something", () => {
  const history = createMemoryHistory()

  const mockClient = {
    executeMutation: jest.fn(),
  }

  return render(
    <Provider value={mockClient}>
      <HelmetProvider>
        <Router history={history}><SomeComponent/></Router>
      </HelmetProvider>
    </Provider>
  )
}

```

urql version & exchanges: `^2.0.5`, no exchanges

**Comment by kitten** [comment, marked as answer]
@basvandriel Hiya 👋 I converted this to a discussion since it's more of a question rather than an issue.

When you're mocking the `Client` and are passing it on as such, you're never mocking a whole `Client` and are instead selectively implementing mocks for individual methods. This means that type checks cannot pass. You could of course create a client and mock and spy on individual methods, this would however defeat the point of the mock, since you may then accidentally use a real `Client` if you're not careful.

Instead you have to type cast the mocked client to `any` (or additionally to `Client` after) to get TypeScript to recognise it as this type forcefully.

**Reply by basvandriel** [reply]
Thankyou!

```
const render = (ui: React.ReactElement) => {
  const history = createMemoryHistory()

  return mount(
    <Provider
      value={
        {
          executeMutation: jest.fn(() => never),
        } as any
      }
    >
      <Router history={history}>{ui}</Router>
    </Provider>
  )
}
```
